proseco/guide.py

```diff
@@ -153,13 +159,56 @@ def run_search_stages(self):
                 self.log(f'Quitting after stage {idx - 1} with {already_selected} stars', level=1)
                 break
         self.log('Done with search stages')
-        selected = cand_guides[cand_guides['stage'] != -1]
-        selected.sort(['stage', 'mag'])
-        if len(selected) >= self.n_guide:
-            return selected[0:n_guide]
-        if len(selected) < self.n_guide:
+        stage_cands = cand_guides[cand_guides['stage'] != -1]
+        stage_cands.sort(['stage', 'mag'])
+        guides = self.select_catalog(stage_cands[0:n_guide + GUIDE.surplus_stars])
+        if len(guides) < self.n_guide:
             self.log(f'Could not find {self.n_guide} candidates after all search stages')
-            return selected
+        return guides
+
+    def select_catalog(self, stage_cands):
+        """
+        For the candidates selected at any stage (stage-assigned candidates?) select the
+        first combination that satisfies the most number of extra tests.
+        Here the extra tests are just the 3 checks for guide stars that are
+        too tightly clustered.
+        """
+        self.log(f'Selecting catalog from {len(stage_cands)} stage-selected stars')
+
+        def index_combinations(n, m):
+            seen = set()
+            for n_tmp in range(m + 1, n + 1):
+                for comb in combinations(range(n_tmp), m):
+                    if comb not in seen:
+                        seen.add(comb)
+                        yield comb
+
+        # Set a dictionary to save the first combination of that satisfies N tests.
+        # This will be used if no combination satisfies all the tests.
+        select_results = {}
+
+        # I should come back to this and see if the "min" is needed or if the combinations
+        # code runs fine even if we have fewer-than-expected stage_cands to start
+        choose_m = min(len(stage_cands), self.n_guide)
+
+        for n_tries, comb in enumerate(index_combinations(len(stage_cands), choose_m), start=1):
+            cands = stage_cands[list(comb)]  # (note that [(1,2)] is not the same as list((1,2))
+            n_pass, n_tests = run_select_checks(cands)  # This function knows how many tests get run
+            if n_pass == n_tests:
+                self.log(f'Selected stars passing all tests at combination {n_tries}',
+                         tried_combinations=n_tries > 1)
```

Comment:
Is the enumerate use here and the new/arbitrary tried_combinations key somewhat sensible to you?

Reply:
Why not just put in `n_tries`.  Then downstream code can do what it wants with that, including testing for `n_tries > 1` or `n_tries > 3` or whatever.

Reply:
The enumerate is good.

Reply:
OK, I went back and forth on just n_tries as well because I wasn't sure what the value should really be for the block at the bottom when n_pass != n_tests .  But I think the last value is reasonable.

Reply:
Also trying to make sure that there's no case where 

index_combinations(len(stage_cands), choose_m)

would error or have no combinations.

Reply:
Ah, so yeah, for index_combinations I'm in a little trouble in small sets of candidates, where I expected N choose M to just give me N if M == N .

In [12]: for comb in combinations([1, 2, 3, 4], 4):
    ...:     print(comb)
    ...:     
(1, 2, 3, 4)

In [13]: for comb in index_combinations(4, 4):
    ...:     print(comb)
    ...: